lib/stm32/common/rtc_common_l1f024.c

```diff
@@ -122,4 +122,189 @@ void rtc_clear_wakeup_flag(void)
+*/
+void rtc_enable_bypass_shadow_register(void)
+{
+	RTC_CR |= RTC_CR_BYPSHAD;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Clears the bypass shadow bit in RTC_CR
+
+@details Requires unlocking backup domain write protection (PWR_CR_DBP)
+*/
+void rtc_disable_bypass_shadow_register(void)
+{
+	RTC_CR &= ~RTC_CR_BYPSHAD;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Sets the RTC hour format (24h or 12h)
+
+@details Requires unlocking backup domain write protection (PWR_CR_DBP)
+*/
+void rtc_set_hour_format(enum rtc_fmt fmt)
+{
+	switch (fmt) {
+	case RTC_CR_FMT_24H:
+		RTC_CR &= ~RTC_CR_FMT;
+		break;
+	case RTC_CR_FMT_12H:
+		RTC_CR |= RTC_CR_FMT;
+		break;
+	}
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Sets the RTC BCD calendar year value (tens and ones)
+
+@details Requires unlocking the RTC write-protection (RTC_WPR)
+*/
+void rtc_calendar_set_year(uint8_t year_tens, uint8_t year_units)
+{
+	RTC_DR = ((year_tens & RTC_DR_YT_MASK) << RTC_DR_YT_SHIFT) |
+		(RTC_DR & ~(RTC_DR_YT_MASK << RTC_DR_YT_SHIFT));
+	RTC_DR = ((year_units & RTC_DR_YU_MASK) << RTC_DR_YU_SHIFT) |
+		(RTC_DR & ~(RTC_DR_YU_MASK << RTC_DR_YU_SHIFT));
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Sets the RTC BCD calendar weekday
+
+@details Requires unlocking the RTC write-protection (RTC_WPR)
+*/
+void rtc_calendar_set_weekday(enum rtc_wdu weekday)
+{
+	RTC_DR = ((weekday & RTC_DR_WDU_MASK) << RTC_DR_WDU_SHIFT) |
+		(RTC_DR & ~(RTC_DR_WDU_MASK << RTC_DR_WDU_SHIFT));
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Sets the RTC BCD calendar month value (tens and ones)
+
+@details Requires unlocking the RTC write-protection (RTC_WPR)
+*/
+void rtc_calendar_set_month(uint8_t month_tens, uint8_t month_units)
+{
+	RTC_DR = ((month_tens & RTC_DR_MT_MASK) << RTC_DR_MT_SHIFT) |
+		(RTC_DR & ~(RTC_DR_MT_MASK << RTC_DR_MT_SHIFT));
+	RTC_DR = ((month_units & RTC_DR_MU_MASK) << RTC_DR_MU_SHIFT) |
+		(RTC_DR & ~(RTC_DR_MU_MASK << RTC_DR_MU_SHIFT));
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Sets the RTC BCD calendar day value (tens and ones)
+
+@details Requires unlocking the RTC write-protection (RTC_WPR)
+*/
+void rtc_calendar_set_day(uint8_t day_tens, uint8_t day_units)
+{
+	RTC_DR = ((day_tens & RTC_DR_DT_MASK) << RTC_DR_DT_SHIFT) |
+		(RTC_DR & ~(RTC_DR_DT_MASK << RTC_DR_DT_SHIFT));
+	RTC_DR = ((day_units & RTC_DR_DU_MASK) << RTC_DR_DU_SHIFT) |
+		(RTC_DR & ~(RTC_DR_DU_MASK << RTC_DR_DU_SHIFT));
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Sets RTC time to use AM or 24-hour format
+
+@details Requires unlocking the RTC write-protection (RTC_WPR)
+*/
+void rtc_time_set_am_notation(void)
+{
+	RTC_TR &= ~RTC_TR_PM;
```

Comment:
I don't believe this does what you think.

As I understand it, the AM/PM bit in the TR register is for the _time_, so, say you want to set the time to 2pm, you initialise by setting hours to 2, and am/pm bit set.

that _loads_ the time, but if you have the CR bit set to 24 hour mode, you would re-read it as hours == 14.

For plain reading, this is just the display bit.  I think it's only writeable here for init time, not a mode change.  I'd probably just make the setting/clearing of this bit an extra boolean parameter on rtc_time_set_hour() ?

Reply:
Combining it to set-hour is a great idea!